scripts/smoketest.ts

```diff
@@ -0,0 +1,397 @@
+import * as execa from 'execa';
+import * as path from 'path';
+import * as fs from 'fs-extra';
+import * as os from 'os';
+import * as pty from 'node-pty';
+import yargs from 'yargs';
+
+export const NPM = {
+  install(pkgName: string, isGlobal: boolean = false) {
+    const args = ['install', pkgName];
+    if (isGlobal) {
+      args.push('-g');
+    }
+    execa.sync('npm', args, { stdio: 'inherit' });
+  },
+  uninstall(pkgName: string, isGlobal: boolean = false) {
+    const args = ['uninstall', pkgName];
+    if (isGlobal) {
+      args.push('-g');
+    }
+    execa.sync('npm', args, { stdio: 'inherit' });
+  },
+};
+
+export type SmoketestArgs = {
+  projectDirectory: string;
+  cliVersion: string;
+  destructive: boolean;
+};
+
+async function getArgs(): Promise<SmoketestArgs> {
+  const args = await yargs(process.argv.slice(2))
+    .option('destructive', {
+      type: 'boolean',
+      description: 'DANGEROUS: Deletes the project directory if it exists',
+      default: false,
+      demandOption: false,
+    })
+    .option('cli-version', {
+      type: 'string',
+      description: 'The version of @aws-amplify/cli to test',
+      demandOption: true,
+    })
+    .option('project-directory', {
+      type: 'string',
+      default: path.join(os.tmpdir(), 'smoketest'),
+    })
+    .help()
+    .alias('help', 'h').argv;
+  return {
+    projectDirectory: args['project-directory'],
+    cliVersion: args['cli-version'],
+    destructive: args.destructive,
+  };
+}
+
+export function assertEmpty(directory: string) {
+  const contents = fs.readdirSync(directory);
+  if (contents.length > 0) {
+    throw new Error(`Project directory ${directory} is not empty`);
+  }
+}
+
+export function createProjectDirectory(projectDirectory: string) {
+  fs.ensureDirSync(projectDirectory);
+}
+
+export function removeProjectDirectory(projectDirectory: string) {
+  fs.removeSync(projectDirectory);
+}
+
+export type Evaluatable = {
+  evaluate(data: string, ptyProcess: pty.IPty): boolean;
+};
```

Comment:
This (and this file) looks a lot like nexpect and our e2e tests.
I'm wondering if we could just throw couple of e2e tests after `deploy` step in release/release_rc pipelines but use amplify installed from npm.

Reply:
Perhaps nexpect should be extracted from e2e core OR this should become lightweight replacement for nexpect.

Reply:
Yeah, the basic implementation pattern is from `nexpect`. I wanted to avoid having to tinker with nexpect and lift it out of `-e2e`, so I ended up just doing a lightweight reimplementation. I would personally vote for avoiding extracting `nexpect`, because I think that may be a larger change than it is worth. However, I think turning this into an e2e test is a sane path to take if we want to go that route.

Reply:
We shouldn't have too many nexpect clones though. I'm a bit afraid this will grow organically into e2e core clone. So if we don't like it maybe we should start nexpectv2 outside of e2e core and then adopt it in e2e and smoke tests.

Given size and usefulness I'd say we should merge it. Let's see what others think as well.